src/emqtt.erl

```diff
@@ -1327,7 +1345,18 @@ connected(info, {timeout, TRef, ack}, State = #state{ack_timer     = TRef,
                            pending_calls = timeout_calls(Timeout, Calls)},
     {keep_state, ensure_ack_timer(NewState)};
 
-connected(info, immediate_retry, State = #state{clean_start = false, inflight = Inflight}) ->
+connected(info, immediate_retry, State = #state{clean_start = false}) ->
+    %% NOTE
+    %% Dropping *random* inflight messages on the floor if the inflight window now
+    %% contains more messages than the current effective max inflight size. This
+    %% is incorrect, strictly speaking, because some of the dropped messages may
```

Comment:
Agree that this is incorrect, but I am not quite following why do we have to drop overflow ?

Reply:
Because we can not retry sending them without hitting reduced `Receive-Maximum` advertised by the server. What do you think are the alternatives?

Reply:
For now, I think we don't really need to handle it by dropping overflow.

Because server may not see the overflow due to lower layer flow limit take effects. 

So two conditions:

1. Server don't see overflow, everyone happy.
2. Server see overflow, it replies PUBACK with RC code, and then we know.  emqtt should not resend it because of ordering.  But it is ok just report back to the caller, so the caller knows by callback. That is to my understanding the same as emqtt drop it with `drop_overflow` and caller get callback ?

Reply:
I agree with you to a high degree. But that would compromise this PR a bit in its attempt to make `emqtt` follow the spec more strictly (that is, if we agree that the spec allows the server to lower `Receive-Maximum` after reconnect for clients with existing sessions). Though I can't really make a good judgement if it's worth to pursue that or not in such a weird corner case.

> Server see overflow, it replies PUBACK with RC code

The worst thing that can happen (as permitted by the spec) is the Server will decide to disconnect us. Then I guess things depend on if the Server drops our session as well.
* If it does, that IIUC the inflight is lost.
* If it doesn't, we again (worst case) risk spinning in a reconnect loop where we resend all the inflight and then get disconnected again.

> That is to my understanding the same as emqtt drop it with drop_overflow and caller get callback ?

Almost. With getting a RC > 127 the publisher will know that this `PUBLISH` was definitely rejected, while with `drop_overflow` they can't be really sure what have happened with their `PUBLISH`, it could have actually been accepted by the server.

---

To add a bit of context: it's not really about EMQX, because EMQX does not control inflight size of QoS1 client publishes.

Reply:
> Because we can not retry sending them without hitting reduced Receive-Maximum advertised by the server. What do you think are the alternatives?

Do you mean the messages that are sent ahead before receiving CONNACK ?

Reply:
No, this (`immediate_retry` phase) happens after `CONNACK`.

The scenario is essentially:
1. `emqtt` connects to Server, which advertises `Receive-Maximum = 10`.
2. Things work smoothy, inflight window is empty most of the time, queue does not build up.
3. At some point connection becomes half-broken, which manifests as Server stops sending ACKs, so that publishes first fill inflight full and the start queuing.
4. TCP finally notices that, signals a disconnect.
5. `emqtt` reconnects, Server says that the session is present but suddenly decides that `Receive-Maximum` should be 5 (note that Server haven't actually seen publishes that are still in the `emqtt` inflight window).
6. Current logic is that `emqtt` should resend everything it has in the inflight, yet there are twice as more publishes as Server is willing to process at the same time.

Please tell if I'm missing something.

Reply:
Thanks. This is clear.
So the consequence of client retry sending the full window is that the server might disconnect it and client would get disconnected immediately hence got stuck in a loop ?

Reply:
Right, though admittedly this is a worst case scenario.